// music-metadata/lib/src/vorbis.js
"use strict";
var strtok = require('strtok2');
(function (VorbisPictureType) {
    VorbisPictureType[VorbisPictureType['Other'] = 0] = 'Other';
    VorbisPictureType[VorbisPictureType["32x32 pixels 'file icon' (PNG only)"] = 1] = "32x32 pixels 'file icon' (PNG only)";
    VorbisPictureType[VorbisPictureType['Other file icon'] = 2] = 'Other file icon';
    VorbisPictureType[VorbisPictureType['Cover (front)'] = 3] = 'Cover (front)';
    VorbisPictureType[VorbisPictureType['Cover (back)'] = 4] = 'Cover (back)';
    VorbisPictureType[VorbisPictureType['Leaflet page'] = 5] = 'Leaflet page';
    VorbisPictureType[VorbisPictureType['Media (e.g. lable side of CD)'] = 6] = 'Media (e.g. lable side of CD)';
    VorbisPictureType[VorbisPictureType['Lead artist/lead performer/soloist'] = 7] = 'Lead artist/lead performer/soloist';
    VorbisPictureType[VorbisPictureType['Artist/performer'] = 8] = 'Artist/performer';
    VorbisPictureType[VorbisPictureType['Conductor'] = 9] = 'Conductor';
    VorbisPictureType[VorbisPictureType['Band/Orchestra'] = 10] = 'Band/Orchestra';
    VorbisPictureType[VorbisPictureType['Composer'] = 11] = 'Composer';
    VorbisPictureType[VorbisPictureType['Lyricist/text writer'] = 12] = 'Lyricist/text writer';
    VorbisPictureType[VorbisPictureType['Recording Location'] = 13] = 'Recording Location';
    VorbisPictureType[VorbisPictureType['During recording'] = 14] = 'During recording';
    VorbisPictureType[VorbisPictureType['During performance'] = 15] = 'During performance';
    VorbisPictureType[VorbisPictureType['Movie/video screen capture'] = 16] = 'Movie/video screen capture';
    VorbisPictureType[VorbisPictureType['A bright coloured fish'] = 17] = 'A bright coloured fish';
    VorbisPictureType[VorbisPictureType['Illustration'] = 18] = 'Illustration';
    VorbisPictureType[VorbisPictureType['Band/artist logotype'] = 19] = 'Band/artist logotype';
    VorbisPictureType[VorbisPictureType['Publisher/Studio logotype'] = 20] = 'Publisher/Studio logotype';
})(exports.VorbisPictureType || (exports.VorbisPictureType = {}));
var VorbisPictureType = exports.VorbisPictureType;
var VorbisPictureParser = (function () {
    function VorbisPictureParser(buffer) {
        var offset = 0;
        this.type = VorbisPictureType[strtok.UINT32_BE.get(buffer, 0)];
        var mimeLen = strtok.UINT32_BE.get(buffer, offset += 4);
        this.format = buffer.toString('utf-8', offset += 4, offset + mimeLen);
        var descLen = strtok.UINT32_BE.get(buffer, offset += mimeLen);
        this.description = buffer.toString('utf-8', offset += 4, offset + descLen);
        this.width = strtok.UINT32_BE.get(buffer, offset += descLen);
        this.height = strtok.UINT32_BE.get(buffer, offset += 4);
        this.colour_depth = strtok.UINT32_BE.get(buffer, offset += 4);
        this.indexed_color = strtok.UINT32_BE.get(buffer, offset += 4);
        var picDataLen = strtok.UINT32_BE.get(buffer, offset += 4);
        this.data = new Buffer(buffer.slice(offset += 4, offset + picDataLen));
    }
    return VorbisPictureParser;
}());
var Vorbis = (function () {
    function Vorbis() {
    }
    Vorbis.readPicture = function (buffer) {
        return new VorbisPictureParser(buffer);
    };
    Vorbis.getPictureType = function (type) {
        return VorbisPictureType[type];
    };
    return Vorbis;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Vorbis;
